File: mlx-sys/shim/src/fft.cc
#include "cxx_mlx_shim/fft.h"
#include "cxx_mlx_shim/shim_helpers.h"
#include "mlx/fft.h"
#include <stdexcept>
namespace cxx_mlx {
std::unique_ptr<MlxArray> ops_real_fft(const MlxArray& input, int32_t n, int32_t axis, bool inverse, uint8_t norm,
    bool has_target, bool is_device_only, uint8_t device_type, int32_t stream_index) {
  auto target = helpers::decode_stream_or_device(has_target, is_device_only, device_type, stream_index);
  if (norm > 2) throw std::invalid_argument("invalid FFT normalization");
  auto normalization = norm == 0 ? mlx::core::fft::FFTNorm::Backward :
      norm == 1 ? mlx::core::fft::FFTNorm::Ortho : mlx::core::fft::FFTNorm::Forward;
  return std::make_unique<MlxArray>(inverse ? mlx::core::fft::irfft(input, n, axis, normalization, target)
      : mlx::core::fft::rfft(input, n, axis, normalization, target));
}
}

File: mlx-sys/src/bridge/fft.rs
//! Real Fourier transform bridge.
#[allow(clippy::missing_safety_doc, clippy::too_many_arguments)]
#[cxx::bridge(namespace = "cxx_mlx")]
pub mod ffi {
    unsafe extern "C++" {
        include!("cxx_mlx_shim/fft.h");
        type MlxArray = crate::bridge::array::ffi::MlxArray;
        unsafe fn ops_real_fft(
            input: &MlxArray,
            n: i32,
            axis: i32,
            inverse: bool,
            norm: u8,
            has_target: bool,
            is_device_only: bool,
            device_type: u8,
            stream_index: i32,
        ) -> Result<UniquePtr<MlxArray>>;
    }
}
